File: packages/mobile/android/src/main/kotlin/app/abto/sdk/AbtoConfig.kt
package app.abto.sdk

import java.net.URI

enum class AbtoEnvironment {
    DEVELOPMENT, STAGING, PRODUCTION;

    val wireName: String get() = name.lowercase()
}

class AbtoInitException(message: String) : IllegalArgumentException(message)

class AbtoConfig(
    val projectKey: String,
    endpoint: String? = null,
    val environment: AbtoEnvironment = AbtoEnvironment.PRODUCTION,
    debug: Boolean? = null,
    val batchSize: Int = ABTO_DEFAULT_BATCH_SIZE,
    val flushIntervalMs: Long = ABTO_DEFAULT_FLUSH_INTERVAL_MS,
) {
    val endpoint: String
    val debug: Boolean

    init {
        if (projectKey.isBlank()) {
            throw AbtoInitException(ABTO_ERR_PROJECT_KEY_REQUIRED)
        }
        if (batchSize !in ABTO_MIN_BATCH_SIZE..ABTO_MAX_BATCH_SIZE) {
            throw AbtoInitException(ABTO_ERR_BATCH_SIZE_RANGE)
        }
        val raw = endpoint ?: ABTO_DEFAULT_COLLECT_ENDPOINT
        // Allow only HTTP(S) endpoints, matching Browser SDK validation.
        val parsed = try {
            URI(raw)
        } catch (_: Exception) {
            null
        }
        val scheme = parsed?.scheme?.lowercase()
        val host = parsed?.host?.lowercase()
        if ((scheme != "http" && scheme != "https") || host.isNullOrBlank()) {
            throw AbtoInitException("$ABTO_ERR_ENDPOINT_INVALID_PREFIX\"$raw\"")
        }
        val developmentLoopback =
            environment == AbtoEnvironment.DEVELOPMENT &&
                (host in ABTO_LOOPBACK_HOSTS || ABTO_LOOPBACK_HOST_PREFIXES.any { host.startsWith(it) })
        if (scheme == "http" && !developmentLoopback) {
            throw AbtoInitException(ABTO_ERR_ENDPOINT_HTTPS_REQUIRED)
        }
        this.endpoint = raw
        this.debug = debug ?: (environment == AbtoEnvironment.DEVELOPMENT)
    }
}

File: packages/mobile/swift/Sources/AbtoApp/AbtoConfig.swift
import Foundation

public enum AbtoEnvironment: String {
    case development
    case staging
    case production
}

public enum AbtoInitError: Error, CustomStringConvertible, LocalizedError {
    case invalidConfig(String)

    public var description: String {
        switch self {
        case .invalidConfig(let message): return message
        }
    }

    public var errorDescription: String? { description }
}

public struct AbtoConfig {
    public let projectKey: String
    public let endpoint: URL
    public let environment: AbtoEnvironment
    public let debug: Bool
    public let batchSize: Int
    public let flushInterval: TimeInterval

    public init(
        projectKey: String,
        endpoint: String? = nil,
        environment: AbtoEnvironment = .production,
        debug: Bool? = nil,
        batchSize: Int = abtoDefaultBatchSize,
        flushInterval: TimeInterval = abtoDefaultFlushInterval
    ) throws {
        guard !projectKey.trimmingCharacters(in: .whitespaces).isEmpty else {
            throw AbtoInitError.invalidConfig(abtoErrProjectKeyRequired)
        }
        guard (abtoMinBatchSize...abtoMaxBatchSize).contains(batchSize) else {
            throw AbtoInitError.invalidConfig(abtoErrBatchSizeRange)
        }
        let rawEndpoint = endpoint ?? abtoDefaultCollectEndpoint
        // Allow only HTTP(S) endpoints, matching Browser SDK validation.
        guard let url = URL(string: rawEndpoint),
              let scheme = url.scheme?.lowercased(),
              scheme == "http" || scheme == "https",
              url.host != nil,
              url.user == nil,
              url.password == nil
        else {
            throw AbtoInitError.invalidConfig("\(abtoErrEndpointInvalidPrefix)\"\(rawEndpoint)\"")
        }
        let host = url.host?.lowercased()
        let developmentLoopback = environment == .development
            && (abtoLoopbackHosts.contains(host ?? "") || abtoLoopbackHostPrefixes.contains { host?.hasPrefix($0) == true })
        guard scheme == "https" || developmentLoopback else {
            throw AbtoInitError.invalidConfig(abtoErrEndpointHTTPSRequired)
        }
        self.projectKey = projectKey
        self.endpoint = url
        self.environment = environment
        self.debug = debug ?? (environment == .development)
        self.batchSize = batchSize
        self.flushInterval = flushInterval
    }
}
